docs/optimization.md

```diff
@@ -0,0 +1,143 @@
+  ; do even things
+)
+```
+*Note that this technique is handy when recursing through lists too.  The last item in a list is always () which evaluates to false, so in that case you can break the recursion.*
+
+However, `divmod` is a pretty expensive operation, and we have to add an `r` to access the remainder once the operation has completed.  Instead, we can just use `logand` to evaluate just the last bit:
+
+```chialisp
+(if (logand value 1)
+  ; do odd things
+  ; do even things
+)
+```
+
+We have now saved ourselves at least 50% of the cost of this block!
+
+## Keep argument numbers small
+
+This tip is both good for optimization and readability.  As the program is running, it needs to pay cost to look up a value from the environment.  It is not a large cost, but it gets larger the deeper it has to go into the environment tree to search for the value.  If you can keep the argument numbers small, you can trim off cost every time your program uses an argument in its evaluation.
+
+One way to do this is to batch arguments that always end up in the same place together.  Here's an example:
+
+```chialisp
+(mod (
+      CURRIED_PUBKEY
+      some_data
+      some_other_data
+      some_more_data
+      even_more_data
+      pubkey
+      my_amount
+      my_id
+     )
+
+     (import "condition_codes.clvm")
+     (import "sha256tree.clvm")
+
+     (defun-inline agg_sig (CURRIED_PUBKEY some_data some_other_data some_more_data even_more_data)
+        (AGG_SIG_ME CURRIED_PUBKEY (sha256tree (list some_data some_other_data some_more_data even_more_data)))
+     )
+
+     (defun-inline assert_amount_and_sig (CURRIED_PUBKEY some_data some_other_data some_more_data even_more_data my_amount)
+        (c (ASSERT_MY_AMOUNT my_amount) (agg_sig CURRIED_PUBKEY some_data some_other_data some_more_data even_more_data))
+     )
+
+     (defun-inline assert_id_and_amount_and_sig (CURRIED_PUBKEY some_data some_other_data some_more_data even_more_data my_amount my_id)
+        (c ASSERT_MY_ID my_id (assert_amount_and_sig CURRIED_PUBKEY some_data some_other_data some_more_data even_more_data my_amount))
+     )
+
+     (assert_id_and_amount_and_sig CURRIED_PUBKEY some_data some_other_data some_more_data even_more_data my_amount my_id)
+)
+```
+
+You can see the code is a little bit out of control from a readability standpoint, and when accessing `my_amount` and `my_id` we have to go deep into the environment tree to read their values.  Instead, we should just batch all of our data into a list to start with.
+
+```chialisp
+(mod (
+      CURRIED_PUBKEY
+      all_data
+      pubkey
+      my_amount
+      my_id
+     )
+
+     (import "condition_codes.clvm")
+     (import "sha256tree.clvm")
+
+     (defun-inline agg_sig (CURRIED_PUBKEY all_data)
+        (AGG_SIG_ME CURRIED_PUBKEY (sha256tree all_data))
+     )
+
+     (defun-inline assert_amount_and_sig (CURRIED_PUBKEY all_data my_amount)
+        (c (ASSERT_MY_AMOUNT my_amount) (agg_sig CURRIED_PUBKEY all_data))
```

Comment:
Well, except `all_data` is a cons box and you can't pass that to `AGG_SIG`.

Reply:
It gets tree hashed in the function above yeah?

Reply:
Ah, got it.